From bdf1fae957ac4b19330b712fd1d09bcb33375be0 Mon Sep 17 00:00:00 2001
From: Varun Sampat <vsampat@twitter.com>
Date: Sat, 30 Nov 2019 17:11:59 -0500
Subject: [PATCH] FRU state sensor for power cycle SEL

---
 include/EventSensor.hpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/include/EventSensor.hpp b/include/EventSensor.hpp
index 1a12d42..2cca8d1 100644
--- a/include/EventSensor.hpp
+++ b/include/EventSensor.hpp
@@ -38,7 +38,8 @@ struct sensorTypeMap
 static const std::vector<struct sensorTypeMap> sensorList = {
     {.sensorName = "ACPI_Power_State", .sensorType = "acpi/"},
     {.sensorName = "Button", .sensorType = "button/"},
-    {.sensorName = "End_of_Post", .sensorType = "systemevent/"}
+    {.sensorName = "End_of_Post", .sensorType = "systemevent/"},
+    {.sensorName = "IPMI_POWER_CYCLE", .sensorType = "fru_state/"}
 };
 
 enum class sensorListIdx : uint8_t
