From bb13d43956a5b0f079bd0e6ae73f4c8833a8a1a3 Mon Sep 17 00:00:00 2001
From: Frederick Lee <Frederick_Lee@wiwynn.com>
Date: Wed, 13 Jan 2021 22:31:59 +0800
Subject: [PATCH] Free ASD MSG memory after running asd_msg_free function

---
 asd_main.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/asd_main.c b/asd_main.c
index 3fde728..abaa6d2 100644
--- a/asd_main.c
+++ b/asd_main.c
@@ -298,6 +298,8 @@ void deinit_asd_state(asd_state* state)
     {
         asd_msg_free(state->asd_msg);
     }
+    free(state->asd_msg);
+    state->asd_msg = NULL;
 }
 
 STATUS send_out_msg_on_socket(void* state, unsigned char* buffer, size_t length)
@@ -799,6 +801,8 @@ STATUS on_client_disconnect(asd_state* state)
                     ASD_LogOption_None, "Failed to de-initialize the asd_msg");
             result = ST_ERR;
         }
+        free(state->asd_msg);
+        state->asd_msg = NULL;
     }
 
     if (result == ST_OK)
-- 
2.17.1

