From ef07a68e7f5d8763ee8b135aee003145ef6db254 Mon Sep 17 00:00:00 2001
From: Frederick Lee <Frederick_Lee@wiwynn.com>
Date: Thu, 19 Nov 2020 17:03:05 +0800
Subject: [PATCH 2/2] Set the tap reset to force when going to idle state

---
 jtag_handler.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/jtag_handler.c b/jtag_handler.c
index 7d176b0..f13e3ff 100644
--- a/jtag_handler.c
+++ b/jtag_handler.c
@@ -241,8 +241,18 @@ STATUS JTAG_set_tap_state(JTAG_Handler* state, enum jtag_states tap_state)
     struct jtag_tap_state tap_state_t;
     // tap_state_t.from = state->active_chain->tap_state;
     tap_state_t.endstate = tap_state;
+/*
     tap_state_t.reset =
         (tap_state == jtag_tlr) ? JTAG_FORCE_RESET : JTAG_NO_RESET;
+*/
+    if ((tap_state == jtag_tlr) || (tap_state == jtag_rti))
+    {
+        tap_state_t.reset = JTAG_FORCE_RESET;
+    }
+    else
+    {
+        tap_state_t.reset = JTAG_NO_RESET;
+    }
     tap_state_t.tck = 1;
 #endif
 
@@ -526,6 +536,7 @@ STATUS JTAG_set_active_chain(JTAG_Handler* state, scanChain chain)
     }
 
     state->active_chain = &state->chains[chain];
+    JTAG_set_tap_state(state, jtag_rti);
 
     return ST_OK;
 }
-- 
2.17.1

