From 55d36f089915354fd61d2891009aa2e2ab0a9cc1 Mon Sep 17 00:00:00 2001
From: Frederick Lee <Frederick_Lee@wiwynn.com>
Date: Tue, 19 May 2020 16:20:50 +0800
Subject: [PATCH] Add to support IPMI mc warm reset command

---
 globalhandler.cpp | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/globalhandler.cpp b/globalhandler.cpp
index c192cbb..2f10244 100644
--- a/globalhandler.cpp
+++ b/globalhandler.cpp
@@ -59,5 +59,10 @@ void register_netfn_global_functions()
     ipmi::registerHandler(ipmi::prioOpenBmcBase, ipmi::netFnApp,
                           ipmi::app::cmdColdReset, ipmi::Privilege::Admin,
                           ipmiGlobalReset);
+
+    // Warm Reset
+    ipmi::registerHandler(ipmi::prioOpenBmcBase, ipmi::netFnApp,
+                          ipmi::app::cmdWarmReset, ipmi::Privilege::Admin,
+                          ipmiGlobalReset);
     return;
 }
-- 
2.24.1

