From c84f2ba9f26d7732318337b8d6f7f166f56d33f5 Mon Sep 17 00:00:00 2001
From: Ray Lue <ray.lue@mic.com.tw>
Date: Thu, 17 Sep 2020 19:27:15 +0800
Subject: [PATCH 1/1] workaround to eliminate power on error

---
 chassishandler.cpp | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/chassishandler.cpp b/chassishandler.cpp
index 269a690..f0569cb 100755
--- a/chassishandler.cpp
+++ b/chassishandler.cpp
@@ -1232,7 +1232,6 @@ ipmi::RspType<> ipmiChassisControl(uint8_t chassisControl)
     static constexpr char const* fruStateSensorPath= "/xyz/openbmc_project/sensors/fru_state/IPMI_Power_Diag";
     std::vector<uint8_t> powerControlEventData{0x02,0x01,0xff};
     int rc = 0;
-
     switch (chassisControl)
     {
         case CMD_POWER_ON:
@@ -1267,6 +1266,14 @@ ipmi::RspType<> ipmiChassisControl(uint8_t chassisControl)
                 // Now request the shutdown
                 generateSELEvent(fruStatePowerOff,powerControlEventData);
                 rc = initiate_state_transition(State::Host::Transition::Off);
+#if 1           // handle the special case: when power is off, and "ipmitool power off;power on" immediately
+                // the power on target will be stopped by pervios power off targets chain.
+                // this delay pervent this error.
+                std::optional<bool> powerGood = power_policy::getPowerStatus();
+                if(powerGood && !(*powerGood)){
+                    std::this_thread::sleep_for (std::chrono::seconds(2));   
+                }
+#endif                 
             }
             else
             {
-- 
2.26.2

