From 2794e1824ab6ccaee9ade12955395b294c23e3ef Mon Sep 17 00:00:00 2001
From: John Chung <john.chung@mic.com.tw>
Date: Thu, 4 Mar 2021 20:21:40 +0800
Subject: [PATCH] Support system event sensor type

---
 include/sdrutils.hpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/sdrutils.hpp b/include/sdrutils.hpp
index 3496c12..c05fc6d 100755
--- a/include/sdrutils.hpp
+++ b/include/sdrutils.hpp
@@ -165,6 +165,7 @@ enum class SensorTypeCodes : uint8_t
     power_supply = 0x08,
     power_unit = 0x09,
     event_disabled = 0x10,
+    system_event = 0x12,
     pwr_button = 0x14,
     restart = 0x1d,
     acpi = 0x22,
@@ -185,6 +186,7 @@ const static boost::container::flat_map<const char*, SensorTypeCodes, CmpStr>
                  {"power_supply", SensorTypeCodes::power_supply},
                  {"power_unit", SensorTypeCodes::power_unit},
                  {"event_disabled", SensorTypeCodes::event_disabled},
+                 {"system_event", SensorTypeCodes::system_event},
                  {"pwr_button", SensorTypeCodes::pwr_button},   
                  {"restart", SensorTypeCodes::restart},  
                  {"acpi", SensorTypeCodes::acpi},     
-- 
2.7.4

