diff --git a/drivers/nvmem/core.c b/drivers/nvmem/core.c
index 3ce6592..48fb03c 100644
--- a/drivers/nvmem/core.c
+++ b/drivers/nvmem/core.c
@@ -435,10 +435,29 @@ static int nvmem_setup_compat(struct nvmem_device *nvmem,
 	if (!config->base_dev)
 		return -EINVAL;
 
+#if 1 // IAN_CHOU
+	if (nvmem->read_only) {
+		if (config->root_only) {
+			nvmem->eeprom = bin_attr_ro_root_nvmem;
+		}
+		else {
+			nvmem->eeprom = bin_attr_ro_nvmem;
+		}
+	}
+	else {
+		if (config->root_only) {
+			nvmem->eeprom = bin_attr_rw_root_nvmem;
+		}
+		else {
+			nvmem->eeprom = bin_attr_rw_nvmem;
+		}
+	}
+#else
 	if (nvmem->read_only)
 		nvmem->eeprom = bin_attr_ro_root_nvmem;
 	else
 		nvmem->eeprom = bin_attr_rw_root_nvmem;
+#endif
 	nvmem->eeprom.attr.name = "eeprom";
 	nvmem->eeprom.size = nvmem->size;
 #ifdef CONFIG_DEBUG_LOCK_ALLOC
